qt/python/mantidqt/utils/writetosignal.py

```diff
@@ -22,8 +22,8 @@ class WriteToSignal(QObject):
 
     def __init__(self, original_out):
         QObject.__init__(self)
-        # If the file descriptor of the stream is < 0 then we are running in a no-external-console mode
-        if not hasattr(original_out, 'fileno') or original_out.fileno() < 0:
+        # Check if the output stream is a printable terminal.
+        if not original_out.isatty():
```

Comment:
What are we actually trying to test here?

I get the feeling that it's "does this go somewhere useful",  which I don't think we have enough information to determine.

For example, stdout redirected to a file is no longer aTTY, but if you are redirecting it you almost certainly want it to output.

Reply:
Output from scripts doesn't seem to be sent to files even on master, unless I'm doing the command wrong (`./bin/launch_mantidworkbench.sh &> ~/mantidoutput.txt`), and the output is still sent to the file even with this change.

Reply:
Ah, maybe the launch scripts have some affects here. I also haven't checked where `WriteToSignal` is actually used.

However looking at the original issue I'm not sure what this fix is supposed to do or what the exception you are trying to remove is.
If it is the unicode error you can test `sys.stdout.encoding` and see if it begins with `utf`. That isn't an ideal test but will be good enough for us and is definitely better than testing if it is a TTY (a TTY may or may not support unicode, similarly for a regular file).

Reply:
The fix is mostly to stop sending to a hidden stdout that doesn't support unicode when starting the program from the button in the "Show Applications" menu on Linux and MacOS.

The exception being caught was added just before release in the `except UnicodeEncodeError`. All this fix does is stop the attempt to output and wasting time if the stdout cannot support unicode (which I assumed all TTY terminals could handle). Thinking about it, using that encoding check would probably work better.

Reply:
@DanNixon The TTY check currently seems more reliable than the unicode check. Do you think this is worth doing at all?

Reply:
Yes, but I'm honestly not sure there is any value in this PR at this point. Does this actually address an issue or is it just replacing a broken hacky "fix" with another broken hacky "fix".

Reply:
It is arguably a better "fix" than what is there previously, as you avoid having to catch an exception every time something that is output to the terminal contains a Unicode character.

Reply:
> as you avoid having to catch an exception every time something that is output to the terminal contains a Unicode character

No you don't. That check needs to remain.

Reply:
Ah, I meant that the catching occurs less often. Which, as I understand it, has some overhead that is avoided by doing the check on the terminal earlier in the code.

Reply:
Are you talking about catching `UnicodeError` in `write()`?

If so then no, that still needs st stay.
Being a TTY reveals zero information about Unicode support.